File: Blind75-Practice/easy/09-heap-priority-queue/Python/kth-largest-element-stream.py
"""
LeetCode 703: Kth Largest Element in a Stream
Difficulty: Easy
Pattern: Heap / Priority Queue

Problem:
Design a class to find the kth largest element in a stream. Note that it is the
kth largest element in the sorted order, not the kth distinct element.

Implement KthLargest class:
- KthLargest(int k, int[] nums) Initializes the object with the integer k and the stream of integers nums.
- int add(int val) Appends the integer val to the stream and returns the element representing the kth largest element in the stream.

Example 1:
Input:
["KthLargest", "add", "add", "add", "add", "add"]
[[3, [4, 5, 8, 2]], [3], [5], [10], [9], [4]]
Output:
[null, 4, 5, 5, 8, 8]

Constraints:
- 1 <= k <= 10^4
- 0 <= nums.length <= 10^4
- -10^4 <= nums[i] <= 10^4
- -10^4 <= val <= 10^4
- At most 10^4 calls will be made to add

Time Complexity: O(log k) per add operation
Space Complexity: O(k)
"""

import heapq
from typing import List


class KthLargest:
    """
    Approach: Use min heap of size k

    Strategy:
    1. Maintain a min heap of the k largest elements
    2. The root of the heap (smallest in heap) is kth largest overall
    3. When adding new element:
       - If heap size < k, add element
       - If new element > heap root, remove root and add new element
    4. Return heap root (kth largest)
    """

    def __init__(self, k: int, nums: List[int]):
        self.k = k
        self.heap = []

        # Add all initial elements
        for num in nums:
            self.add(num)

    def add(self, val: int) -> int:
        # If heap not full, add element
        if len(self.heap) < self.k:
            heapq.heappush(self.heap, val)
        # If new element larger than smallest in heap, replace it
        elif val > self.heap[0]:
            heapq.heapreplace(self.heap, val)

        # Return kth largest (root of min heap)
        return self.heap[0]


# Alternative implementation with cleaner logic
class KthLargestAlternative:
    def __init__(self, k: int, nums: List[int]):
        self.k = k
        self.heap = nums
        heapq.heapify(self.heap)

        # Keep only k largest elements
        while len(self.heap) > k:
            heapq.heappop(self.heap)

    def add(self, val: int) -> int:
        heapq.heappush(self.heap, val)

        if len(self.heap) > self.k:
            heapq.heappop(self.heap)

        return self.heap[0]


# Test cases
if __name__ == "__main__":
    # Test case 1
    kthLargest = KthLargest(3, [4, 5, 8, 2])
    assert kthLargest.add(3) == 4  # return 4
    print("✓ Test 1 passed")

    assert kthLargest.add(5) == 5  # return 5
    print("✓ Test 2 passed")

    assert kthLargest.add(10) == 5  # return 5
    print("✓ Test 3 passed")

    assert kthLargest.add(9) == 8  # return 8
    print("✓ Test 4 passed")

    assert kthLargest.add(4) == 8  # return 8
    print("✓ Test 5 passed")

    print("\nAll test cases passed!")

File: Blind75-Practice/easy/09-heap-priority-queue/Java/KthLargestElementStream.java
/*
 * 703. Kth Largest Element in a Stream
 * Difficulty: Easy
 * Pattern: Heap
 *
 * Problem:
 * Design a class that always knows the kth largest value seen so far in a stream.
 * You are given an integer `k` and an initial list of numbers `nums`. Each time a
 * new value is added, return the current kth largest element among all values seen.
 *
 * Example 1:
 * Input:
 *   k = 3
 *   nums = [4, 5, 8, 2]
 *   add(3), add(5), add(10), add(9), add(4)
 * Output:
 *   4, 5, 5, 8, 8
 * Explanation:
 * - Stream after add(3):  [4, 5, 8, 2, 3]  -> 3rd largest = 4
 * - Stream after add(5):  [4, 5, 8, 2, 3, 5] -> 3rd largest = 5
 * - Stream after add(10): [4, 5, 8, 2, 3, 5, 10] -> 3rd largest = 5
 * - Stream after add(9):  [4, 5, 8, 2, 3, 5, 10, 9] -> 3rd largest = 8
 * - Stream after add(4):  [4, 5, 8, 2, 3, 5, 10, 9, 4] -> 3rd largest = 8
 *
 * Example 2:
 * Input:
 *   k = 1
 *   nums = []
 *   add(-3), add(-2), add(-4), add(0), add(4)
 * Output:
 *   -3, -2, -2, 0, 4
 * Explanation:
 * When k = 1, we always return the largest value seen so far.
 *
 * Constraints:
 * - 1 <= k <= 10^4
 * - 0 <= nums.length <= 10^4
 * - -10^4 <= nums[i], val <= 10^4
 * - At most 10^4 calls will be made to `add`
 *
 * Why a heap works:
 * Keep only the k largest values in a min-heap. The heap root is then the
 * smallest of those k values, which is exactly the kth largest overall.
 *
 * Time: O(log k) per add
 * Space: O(k)
 */

import java.util.PriorityQueue;

public class KthLargestElementStream {

    static class KthLargest {
        private final int k;
        private final PriorityQueue<Integer> minHeap = new PriorityQueue<>();

        public KthLargest(int k, int[] nums) {
            this.k = k;
            for (int num : nums) {
                add(num);
            }
        }

        // Single place for heap logic: offer, trim, return kth largest.
        public int add(int val) {
            minHeap.offer(val);
            if (minHeap.size() > k) {
                minHeap.poll(); // removes smallest (heap root), not a random leaf
            }
            return minHeap.peek();
        }
    }
            //

            //

            //

            //

            //

            //

            //

            //

            //

            //

            //

            //

            //

            //

            //

            //

            //

            //

            //

            //

            //

            //

            //

            //

            //

            //

            //

            //

            //

            //

            //

            //

            //

            //

            //

            //

            //

            //

            //

            //

            //

            //

            //

            //

            //

            //

            //

            //

            //

            //

            // public KthLargest(int k, int[] nums) {
            //     this.k = k;
            //     for (int num : nums) {
            //         add(num);
            //     }
            // }
            //
            // public int add(int val) {
            //     minHeap.offer(val);
            //     if (minHeap.size() > k) {
            //         minHeap.poll();
            //     }
            //     return minHeap.peek();
            // }

    public static void main(String[] args) {
        KthLargest kth = new KthLargest(3, new int[] {4, 5, 8, 2});
        assert kth.add(3) == 4;
        assert kth.add(5) == 5;
        assert kth.add(10) == 5;
        assert kth.add(9) == 8;
        assert kth.add(4) == 8;
        System.out.println("All test cases passed!");
    }
}
